apps/src/templates/sectionProgress/progressTables/ProgressTableContainer.jsx

```diff
@@ -80,27 +103,109 @@ class ProgressTableContainer extends React.Component {
   // account for the horizontal space used by the vertical scrollbar.
   needsContentHeaderGutter() {
     return (
-      this.props.section.students.length *
-        parseInt(progressTableStyles.ROW_HEIGHT) >
-      parseInt(progressTableStyles.MAX_BODY_HEIGHT)
+      this.props.section.students.length >
+      parseInt(progressTableStyles.MAX_ROWS)
+    );
+  }
+
+  onToggleRow(rowData) {
+    const rowIndex = this.state.rows.findIndex(
+      row => row.student === rowData.student
+    );
+    if (!rowData.isExpanded) {
+      this.expandDetailRows(rowData, rowIndex);
+    } else {
+      this.collapseDetailRows(rowData, rowIndex);
+    }
+    this.recordToggleRow(!rowData.isExpanded, rowData.student.id);
+  }
+
+  recordToggleRow(expanding, studentId) {
+    firehoseClient.putRecord(
+      {
+        study: 'teacher_dashboard_actions',
+        study_group: 'time_spent',
+        event: 'toggle_details',
+        data_json: JSON.stringify({
+          student_id: studentId,
+          section_id: this.props.section.id,
+          visible: expanding
+        })
+      },
+      {includeUserId: true}
     );
   }
 
+  expandDetailRows(rowData, rowIndex) {
+    const detailRows = [];
+    for (let i = 1; i <= this.numDetailRows; i++) {
+      detailRows.push({
+        id: idForExpansionIndex(rowData.student.id, i),
+        student: rowData.student,
+        expansionIndex: i,
+        useDarkBackground: rowData.useDarkBackground
+      });
+    }
+    const rows = [...this.state.rows];
+    rows[rowIndex].isExpanded = true;
+    rows.splice(rowIndex + 1, 0, ...detailRows);
+    this.setState({
+      rows: rows
+    });
+  }
+
+  collapseDetailRows(rowData, rowIndex) {
+    const rows = this.state.rows.filter(row => {
+      return row.student.id !== rowData.student.id || row.expansionIndex === 0;
+    });
+    rows[rowIndex].isExpanded = false;
+    this.setState({
```

Comment:
Fun fact: this could also be written as `this.setState({rows})` since the key and value have the same name

Reply:
👍